plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs

```diff
@@ -154,7 +119,29 @@ fullyApplyAndBetaReduce info args0 = do
               (\n -> if n == param then Just <$> PLC.rename arg else pure Nothing)
               acc
           go acc' arity' args'
-        _ -> pure Nothing
+        -- term/type argument mismatch, don't inline
+        (TermParam _:_, TypeAppContext{}) -> pure Nothing
+        (TypeParam _:_, TermAppContext{}) -> pure Nothing
+        -- no more arguments to apply, just apply what we have
+        (remainingArity, AppContextEnd) ->
+          let numberOfUnappliedLam = length remainingArity
+              allLambdas = fst $ splitLambdas (varRhs info)
+              -- | Returns the function with the unapplied lambda.
+              partiallyAppliedFun ::
+                -- | The function body
+                Term tyname name uni fun ann ->
+                -- | The number of arguments not applied
+                Int ->
+                [Lam tyname name uni ann] -> -- The whole list of lambdas of the function
```

Comment:
This is also not needed - why use the whole list of lambdas? You already have `remainingArity`.

Reply:
I don't really *need* the whole list, but it's what `splitLambdas` returns and it'll be work to filter out the needed ones.

Reply:
If you put additional things you need into `Arity`, then you can simply use `remainingArity`, without needing to `splitLambdas`

Reply:
As an example: `f x y` where `f = \a b c d -> a + b + c + d`.
Here we have `acc = x + y + c + d`, and `remainingArity = [c, d]`. And we need to make `\x y -> x + y + c + d`.
So we can simply add a lambda abstraction to `acc` for each element of `remainingArity`.

Reply:
Yes, I understand, if `Arity` already has all the needed info, and I will make it so.

Reply:
There's `mkIterLamAbs` and `mkIterTyAbs` in `PlutusCore.MkPlc`. Since this function makes both LamAbs and TyAbs, we can perhaps call it `mkAbs`.